tokio-timer/src/timer/entry.rs

```diff
@@ -0,0 +1,559 @@
+///   timer wheel is a head pointer to the list of entries that must be
+///   processed during that timer tick.
+#[derive(Debug)]
+pub struct Entry {
+    /// Timer internals. Using a weak pointer allows the timer to shutdown
+    /// without all `Sleep` instances having completed.
+    inner: Weak<Inner>,
+
+    /// Task to notify once the deadline is reached.
+    task: AtomicTask,
+
+    /// Tracks the entry state. This value contains the following information:
+    ///
+    /// * The deadline at which the entry must be "fired".
+    /// * A flag indicating if the entry has already been fired.
+    /// * Whether or not the entry transitioned to the error state.
+    ///
+    /// When an `Entry` is created, `state` is initialized to the instant at
+    /// which the entry must be fired. When a timer is reset to a different
+    /// instant, this value is changed.
+    state: AtomicU64,
+
+    /// When true, the entry is counted by `Inner` towards the max oustanding
+    /// timeouts. The drop fn uses this to know if it should decrement the
+    /// counter.
+    ///
+    /// One might think that it would be easier to just not create the `Entry`.
+    /// The problem is that `Sleep` expects creating a `Registration` to always
+    /// return a `Registration` instance. This simplifying factor allows it to
+    /// improve the struct layout. To do this, we must always allocate the node.
+    counted: bool,
+
+    /// True wheen the entry is queued in the "process" stack. This value
+    /// is set before pushing the value and unset after popping the value.
+    queued: AtomicBool,
+
+    /// Next entry in the "process" linked list.
+    ///
+    /// Represents a strong Arc ref.
+    next_atomic: UnsafeCell<*mut Entry>,
+
+    /// When the entry expires, relative to the `start` of the timer
+    /// (Inner::start). This is only used by the timer.
+    ///
+    /// A `Sleep` instance can be reset to a different deadline by the thread
+    /// that owns the `Sleep` instance. In this case, the timer thread will not
+    /// immediately know that this has happened. The timer thread must know the
+    /// last deadline that it saw as it uses this value to locate the entry in
+    /// its wheel.
+    ///
+    /// Once the timer thread observes that the instant has changed, it updates
+    /// the wheel and sets this value. The idea is that this value eventually
+    /// converges to the value of `state` as the timer thread makes updates.
+    when: UnsafeCell<Option<u64>>,
+
+    /// Next entry in the State's linked list.
+    ///
+    /// This is only accessed by the timer
+    next_stack: UnsafeCell<Option<Arc<Entry>>>,
+
+    /// Previous entry in the State's linked list.
+    ///
+    /// This is only accessed by the timer and is used to unlink a canceled
+    /// entry.
+    ///
+    /// This is a weak reference.
+    prev_stack: UnsafeCell<*const Entry>,
+}
+
+/// A doubly linked stack
```

Comment:
doubly?

Reply:
It is a stack using a doubly linked list: https://en.wikipedia.org/wiki/Doubly_linked_list